nomos-mix/message/src/layered_cipher.rs

```diff
@@ -0,0 +1,331 @@
+    InvalidDataLength,
+    #[error("Invalid params")]
+    InvalidParams,
+    #[error("Integrity MAC verification failed")]
+    IntegrityMacVerificationFailed,
+}
+
+type Result<T> = std::result::Result<T, Error>;
+
+/// A cipher to encrypt/decrypt a list of data of the same size using a list of keys.
+///
+/// The cipher performs the layered encryption.
+/// The following example shows the simplified output.
+/// - Input: [[data0, k0], [data1, k1]]
+/// - Output: encrypt(k0, [data0, encrypt(k1, [data1])])
+///
+/// The max number of layers is limited to the `max_layers` parameter.
+/// Even if the number of data and keys provided for encryption is smaller than `max_layers`,
+/// The cipher always produces the max-size output regardless of the number of data and keys provided,
+/// in order to ensure that all outputs generated by the cipher are the same size.
+///
+/// The cipher also provides the length-preserved decryption.
+/// Even if one layer of encryptions is decrypted, the length of decrypted data is
+/// the same as the length of the original data.
+/// For example:
+///   len(encrypt(k0, [data0, encrypt(k1, [data1])])) == len(encrypt(k1, [data1]))
+pub struct ConsistentLengthLayeredCipher<D: ConsistentLengthLayeredCipherData> {
+    /// All encrypted data produced by the cipher has the same size according to the `max_layers`.
+    pub max_layers: usize,
+    _phantom_data: PhantomData<D>,
+}
+
+pub trait ConsistentLengthLayeredCipherData {
+    // Returns the serialized bytes for an instance of the implementing type
+    fn to_bytes(&self) -> Vec<u8>;
+    // Returns the size of the serialized data. This is a static method.
+    fn size() -> usize;
+}
+
+/// A parameter for one layer of encryption
+pub struct EncryptionParam<D: ConsistentLengthLayeredCipherData> {
+    /// A data to be included in the layer.
+    pub data: D,
+    /// A [`Key`] to encrypt the layer that will include the [`Self::data`].
+    pub key: Key,
+}
+
+/// A set of keys to encrypt/decrypt a single layer.
+pub struct Key {
+    /// A 128-bit key for encryption/decryption
+    pub stream_cipher_key: StreamCipherKey,
+    /// A 128-bit key for computing/verifying integrity MAC
+    pub integrity_mac_key: HeaderIntegrityMacKey,
+}
+
+impl<D: ConsistentLengthLayeredCipherData> ConsistentLengthLayeredCipher<D> {
+    pub fn new(max_layers: usize) -> Self {
+        Self {
+            max_layers,
+            _phantom_data: Default::default(),
+        }
+    }
+
+    /// The total size of fully encrypted output that includes all layers.
+    /// This size is determined by [`D::size`] and [`Self::max_layers`].
+    pub fn total_size(&self) -> usize {
+        Self::single_layer_size() * self.max_layers
+    }
+
+    /// The size of a single layer that contains a data and a MAC.
+    /// The MAC is used to verify integrity of the encrypted next layer.
+    fn single_layer_size() -> usize {
+        D::size() + HEADER_INTEGRITY_MAC_SIZE
```

Comment:
if `size` method is const, then we can make this method const as well.

Reply:
Here, the data size (`D::size()`) depends on the max size of a block proposal in our case. So, I think it's hard to set a proper constant in this library level.

Reply:
Yeah, but, In our codebase, the `D` is always the `RoutingInfo` struct. If we make it as a constant, we need to update the constant when we add more fields to the struct. And, its size depends on how it is serialized.

Reply:
Thanks. 30eefbb